qiskit_optimization/algorithms/warm_start_qaoa_optimizer.py

```diff
@@ -0,0 +1,339 @@
+
+
+class MeanAggregator(BaseAggregator):
+    """Aggregates the results by averaging the probability of each sample."""
+
+    def aggregate(self, results: List[MinimumEigenOptimizationResult]) \
+            -> List[SolutionSample]:
+        """
+        Args:
+            results: List of result objects that need to be combined.
+
+        Returns:
+             Aggregated samples by averaging them.
+        """
+
+        # Use a dict for fast solution look-up
+        # Key: sample code, value: tuple of fval, probability
+        dict_samples: Dict[str, Tuple[float, float]] = {}
+
+        def _to_string(x: np.ndarray) -> str:
+            return "".join(str(int(v)) for v in x)
+
+        def _from_string(string) -> np.ndarray:
+            return np.array([float(c) for c in string])
+
+        # Sum up all the probabilities in the results
+        for result in results:
+            for sample in result.samples:
+                # state, fval, prob = sample[0], sample[1], sample[2]
+                state, fval, prob = _to_string(sample.x), sample.fval, sample.probability
+                if state in dict_samples:
+                    dict_samples[state] = (fval, dict_samples[state][1] + prob)
+                else:
+                    dict_samples[state] = (fval, prob)
+
+        # Divide by the number of results to normalize
+        aggregated_samples = []
+        num_results = len(results)
+        for state in dict_samples:
+            # sample = (state, dict_samples[state][0], dict_samples[state][1] / num_results)
+            sample = SolutionSample(x=_from_string(state),
+                                    fval=dict_samples[state][0],
+                                    probability=dict_samples[state][1] / num_results,
+                                    status=OptimizationResultStatus.SUCCESS)
+            aggregated_samples.append(sample)
+
+        # new_samples.sort(key=lambda sample: problem.objective.sense.value * sample[1])
+        # x = [float(e) for e in new_samples[0][0]]
+
+        return aggregated_samples
+
+
+class WarmStartQAOACircuitFactory:
+    """
+    A factory that produces quantum circuits for the QAOA implementation. The methods of this
+    factory can be overridden to modify behavior of QAOA. This implementation generates quantum
+    circuits for initial state and mixer to warm start QAOA.
+    """
+
+    def __init__(self, epsilon: float) -> None:
+        """
+        Args:
+            epsilon: the regularization parameter that changes the initial variables according to
+                xi = epsilon if xi < epsilon
+                xi = 1-epsilon if xi > epsilon.
+                The regularization parameter epsilon should be between 0 and 0.5. When it
+                is 0.5 then warm start corresponds to standard QAOA.
+        Raises:
+            AquaError: if ``epsilon`` is not specified for the warm start QAOA or value is not in
+                the range [0, 0.5].
```

Comment:
I do not see it raising AquaError - but just noting that in Optimization repo we should be raising an QiskitOptimizationError now, but of course we may pass on exceptions from Terra unaltered but there Aqua Error is removed too

Reply:
Added a check. Yes, it should be OptimizationError.